### PR TITLE
✨ Add helm plugin to distribute projects

### PR DESCRIPTION
**Pull request by camilamacedo86**
### Add Helm Plugin for Project Distribution

This PR introduces an optional Helm plugin, enabling users to scaffold Helm charts for Kubebuilder projects.

🚀 You can validate the HelmChart generated as example to distribute the project looking at: testdata/project-v4-with-plugins/dist/chart/

### Key Features

- **Helm Chart Scaffolding**: Allows generation of a Helm chart under the `dist/chart` directory, providing an alternative for distributing and managing projects via Helm.
- **Usage**:
  - **To add when init the project**: `kubebuilder init --plugins=helm/v1-alpha` for new projects.
  - **To Init/Update in a project previously scaffolded**: `kubebuilder edit --plugins=helm/v1-alpha` to add or update Helm charts for existing projects.
- **Sync Helm Chart**: Syncs the Helm chart with project manifests using the `edit` command, with `--force` required for updates after webhook or `DeployImage` plugin changes in order to overwritten the values.yaml and manager.yaml files.

### Testing and Documentation

- **Examples**: Usage samples can be found in `testdata/project-v4-with-plugins`.
- **Testing**: Adds GitHub Actions workflows to validate Helm integration, including linting, installation, and e2e checks.

**Note**: For details on the original proposal, see [PR #3632](https://github.com/kubernetes-sigs/kubebuilder/pull/3632).

### CODE REVIEW

File: pkg/plugins/optional/helm/v1alpha/scaffolds/init.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+		DestDir string
+		SubDir  string
+	}{
+		{"config/rbac", "dist/chart/templates/rbac", "rbac"},
+		{"config/webhook", "dist/chart/templates/webhook", "webhook"},
+		{"config/crd/bases", "dist/chart/templates/crds", "crd"},
+		{"config/prometheus", "dist/chart/templates/prometheus", "prometheus"},
+		{"config/certmanager", "dist/chart/templates/certmanager", "certmanager"},
+		{"config/network-policy", "dist/chart/templates/network-policy", "networkPolicy"},
+	}
```
Comment (camilamacedo86) [inline pkg/plugins/optional/helm/v1alpha/scaffolds/init.go:132]: It is just to getting started to implement it.
We probably need move out those to templates
Also, some changes might need to be required to be done in the content to use the labels from the helpers and use the namespace from values.

**Review by camilamacedo86 — COMMENTED** [review]
/hold 

Just to allow people lgtm and approve
so that we can get enough reviews

**Comment by camilamacedo86** [comment]
Hi folks, 

Please feel free to review and give your inputs:

@caiofralmeida, @jgillich, @jrxs-cg, @mhkarimi1383, @Jay-Madden, @maciaszczykm, @hprotzek, @octaviansima, @Cubixmeister, @isac322 @LCaparelli, @dashanji @hprotzek, @Tuhis, @oscartrullols, @sarthaksarthak9, @wille, @algo7, @Mershab99, @pazepaze @caiofralmeida, @hprotzek, @rafalgalaw @zqzten

**Comment by mhkarimi1383** [comment]
Everything looks great 😃 
With that also testing operators will be easier since using helm we can easily provide values

**Review by algo7 — APPROVED** [review]

**Review by mhkarimi1383 — APPROVED** [review]

**Comment by sarthaksarthak9** [comment]
changes are working fine. Everything looks good to me 😊

/lgtm

File: testdata/project-v4-with-plugins/dist/chart/templates/manager/manager.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          {{- toYaml .Values.controllerManager.securityContext | nindent 8 }}
+        volumeMounts:
+        - name: webhook-cert
+          mountPath: /tmp/k8s-webhook-server/serving-certs
+          readOnly: true
+      terminationGracePeriodSeconds: {{ .Values.controllerManager.terminationGracePeriodSeconds }}
+      volumes:
+      - name: webhook-cert
+        secret:
+          secretName: webhook-server-cert
```
Comment (camilamacedo86) [inline testdata/project-v4-with-plugins/dist/chart/templates/manager/manager.yaml:51]: TODO: Change the name for `cert` to be as we scaffold 
Also, add a condition here to only mount when / if cert-manager is enabled

File: docs/book/src/plugins/available/helm-v1-alpha.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
+# Helm Plugin (`helm/v1-alpha`)
```
Comment (camilamacedo86) [inline docs/book/src/plugins/available/helm-v1-alpha.md:1]: TODO: After I further review I discovered an issue
It is not handle the ConversionWebhook approprieted as the cert manager injection for those CRDs

File: testdata/project-v4-with-plugins/dist/chart/templates/manager/manager.yaml (hunk 2 above)
Comment (camilamacedo86) [reply testdata/project-v4-with-plugins/dist/chart/templates/manager/manager.yaml]: done

File: docs/book/src/plugins/available/helm-v1-alpha.md (hunk 3 above)
Comment (camilamacedo86) [reply docs/book/src/plugins/available/helm-v1-alpha.md:1]: Done/fixed

File: .github/workflows/test-helm-samples.yml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          kubectl wait --namespace cert-manager --for=condition=available --timeout=300s deployment/cert-manager-webhook
+
+      - name: Prepare project-v4-with-plugins
+        run: |
+          cd testdata/project-v4-with-plugins/
+          go mod tidy
+          make docker-build IMG=project-v4-with-plugins:v0.1.0
+          kind load docker-image project-v4-with-plugins:v0.1.0
+
+      - name: Install Helm
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply .github/workflows/test-helm-samples.yml:59]: Hi @mrueg 

Thx for your input. That is true. 
However, we are not using this one because it is not promoted by helm itself.
Anyway I think both works and it seems more a matter of preference. 

So, I am closing it as solved but if you think that we must change it please feel free to re-open

File: .github/workflows/test-helm-samples.yml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Setup Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: '~1.22'
+
+      - name: Install the latest version of kind
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply .github/workflows/test-helm-samples.yml:28]: Hi @mrueg 

Thx for your input. That is true. 
However, we are not using this one because it is not promoted by helm itself.
Anyway I think both works and it seems more a matter of preference. 

So, I am closing it as solved but if you think that we must change it please feel free to re-open

File: docs/book/src/plugins/available/helm-v1-alpha.md (hunk 3 above)
Comment (camilamacedo86) [reply docs/book/src/plugins/available/helm-v1-alpha.md:1]: Done

File: pkg/cli/alpha/internal/generate.go [hunk 6]
```diff
@@ -392,3 +399,32 @@ func kubebuilderGrafanaEdit() error {
 	}
 	return nil
 }
+
+// Edits the project to include the Grafana plugin.
```
Comment (monteiro-renato) [inline pkg/cli/alpha/internal/generate.go:403]:
```suggestion
// Edits the project to include the Helm plugin.
```

File: testdata/project-v4-with-plugins/dist/chart/values.yaml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    capabilities:
+      drop:
+      - "ALL"
+  serviceAccountName: project-v4-with-plugins-controller-manager
+  terminationGracePeriodSeconds: 10
+  env:
+    BUSYBOX_IMAGE: busybox:1.36.1
+    MEMCACHED_IMAGE: memcached:1.6.26-alpine3.19
+
+# [CRDs]: To create RBAC (Permissions) configurations
```
Comment (monteiro-renato) [inline testdata/project-v4-with-plugins/dist/chart/values.yaml:42]:
```suggestion
# [RBAC]: To create RBAC (Permissions) configurations
```

File: testdata/project-v4-with-plugins/dist/chart/values.yaml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+# [PROMETHEUS]: To create a ServiceMonitor to export metrics to Prometheus set true
+prometheus:
+  create: false
+
+# [CERT-MANAGER]: To enable cert-manager injection to webhooks set true
+certmanager:
+  create: true
+
+# [NETWORK POLICIES]: To create NetworkPolices set true
```
Comment (monteiro-renato) [inline testdata/project-v4-with-plugins/dist/chart/values.yaml:98]:
```suggestion
# [NETWORK POLICIES]: To create NetworkPolicies set true
```

File: testdata/project-v4-with-plugins/dist/chart/templates/network-policy/allow-metrics-traffic.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+{{- if .Values.networkPolicy.create }}
+# This NetworkPolicy allows ingress traffic
+# with Pods running on namespaces labeled with 'metrics: enabled'. Only Pods on those
+# namespaces are able to gathering data from the metrics endpoint.
```
Comment (monteiro-renato) [inline testdata/project-v4-with-plugins/dist/chart/templates/network-policy/allow-metrics-traffic.yaml:4]:
```suggestion
# namespaces are able to gather data from the metrics endpoint.
```

File: pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/values.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+  serviceAccountName: {{ .ProjectName }}-controller-manager
+  terminationGracePeriodSeconds: 10
+{{- if .DeployImages }}
+  env:
+  {{- range $kind, $image := .DeployImages }}
+    {{ $kind }}_IMAGE: {{ $image }}
+  {{- end }}
+{{- end }}
+
+# [CRDs]: To create RBAC (Permissions) configurations
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/values.go:102]:
```suggestion
# [RBAC]: To create RBAC (Permissions) configurations
```

File: pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/values.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+{{ end }}
+# [PROMETHEUS]: To create a ServiceMonitor to export metrics to Prometheus set true
+prometheus:
+  create: false
+
+# [CERT-MANAGER]: To enable cert-manager injection to webhooks set true
+certmanager:
+  create: {{ .HasWebhooks }}
+
+# [NETWORK POLICIES]: To create NetworkPolices set true
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/values.go:179]:
```suggestion
# [NETWORK POLICIES]: To create NetworkPolicies set true
```

File: .github/workflows/test-helm-samples.yml (hunk 4 above)
Comment (monteiro-renato) [inline .github/workflows/test-helm-samples.yml:59]: Is this expected to be here? helm is already being used above to install cert-manager.
bilde
https://github.com/helm/helm/blob/main/scripts/get-helm-3#L140

File: pkg/plugins/optional/helm/v1alpha/edit.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  - chart/.helmignore
+
+All other files are updated without the usage of the '--force=true' flag 
+when the edit option is used to ensure that the 
+manifests in the chart align with the latest changes.
+`, cliMeta.CommandName)
+}
+
+func (p *editSubcommand) BindFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&p.force, "force", true, "if true, run re-generate the files")
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/edit.go:68]:
```suggestion
	fs.BoolVar(&p.force, "force", true, "if true, regenerates all the files")
```

File: pkg/plugins/optional/helm/v1alpha/init.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+	config config.Config
+}
+
+func (p *initSubcommand) UpdateMetadata(cliMeta plugin.CLIMetadata, subcmdMeta *plugin.SubcommandMetadata) {
+	subcmdMeta.Description = `Initialize a helm chart to distribute the project under dist/
+`
+	subcmdMeta.Examples = fmt.Sprintf(`# Initialize a helm chart to distribute the project under dist/
+  %[1]s init --plugins=helm/v1-alpha
+
+**IMORTANT** You must use %[1]s edit --plugins=helm/v1-alpha to update the chart when changes are made.
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/init.go:40]:
```suggestion
**IMPORTANT** You must use %[1]s edit --plugins=helm/v1-alpha to update the chart when changes are made.
```

File: pkg/plugins/optional/helm/v1alpha/plugin.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	_ plugin.Init = Plugin{}
+	_ plugin.Edit = Plugin{}
+)
+
+type pluginConfig struct{}
+
+// Name returns the name of the plugin
+func (Plugin) Name() string { return pluginName }
+
+// Version returns the version of the grafana plugin
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/plugin.go:51]:
```suggestion
// Version returns the version of the helm plugin
```

File: pkg/plugins/optional/helm/v1alpha/plugin.go [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+func (Plugin) Version() plugin.Version { return pluginVersion }
+
+// SupportedProjectVersions returns an array with all project versions supported by the plugin
+func (Plugin) SupportedProjectVersions() []config.Version { return supportedProjectVersions }
+
+// GetInitSubcommand will return the subcommand which is responsible for initializing and scaffolding grafana manifests
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/plugin.go:57]:
```suggestion
// GetInitSubcommand will return the subcommand which is responsible for initializing and scaffolding a helm chart
```

File: pkg/plugins/optional/helm/v1alpha/plugin.go [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+func (p Plugin) GetInitSubcommand() plugin.InitSubcommand { return &p.initSubcommand }
+
+// GetEditSubcommand will return the subcommand which is responsible for adding grafana manifests
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/plugin.go:60]:
```suggestion
// GetEditSubcommand will return the subcommand which is responsible for editing the helm chart
```
The Grafana comment should probably be updated too, although it can probably be in a different PR.

File: pkg/plugins/optional/helm/v1alpha/scaffolds/init.go [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+	deployImages := make(map[string]string)
+
+	pluginConfig := struct {
+		Resources []struct {
+			Kind    string            `json:"kind"`
+			Options map[string]string `json:"options"`
+		} `json:"resources"`
+	}{}
+
+	const deployImageKey = "deploy-image.go.kubebuilder.io/v1-alpha"
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/scaffolds/init.go:131]: nit: Might be worth it to have [this](https://github.com/kubernetes-sigs/kubebuilder/blob/a9ee3909f7686902879bd666b92deec4718d92c9/pkg/cli/alpha/internal/generate.go#L45-L46) public in order to avoid hard-coding these values in different places.

Comment (camilamacedo86) [reply pkg/plugins/optional/helm/v1alpha/scaffolds/init.go:131]: we would need make public from the plugin itself, not this file.
this file is only for the alpha command
If we could make plugin from the plugin than we can use in this file as well.

Comment (camilamacedo86) [reply pkg/plugins/optional/helm/v1alpha/scaffolds/init.go:131]: It is a good point, but maybe is better to do in a follow up for all plugins

File: .github/workflows/test-helm-samples.yml (hunk 4 above)
Comment (camilamacedo86) [reply .github/workflows/test-helm-samples.yml:59]: good point fixed now.
we need install first then do the rest.

File: pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/chart-templates/manager/manager.go [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+	"sigs.k8s.io/kubebuilder/v4/pkg/machinery"
+)
+
+var _ machinery.Template = &ManagerDeployment{}
+
+// ManagerDeployment scaffolds the manager Deployment for the Helm chart
+type ManagerDeployment struct {
+	machinery.TemplateMixin
+
+	// Force if true will scaffold the env with the images
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/chart-templates/manager/manager.go:31]:
```suggestion
	// DeployImages if true will scaffold the env with the images
```

File: pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/chart.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"path/filepath"
+
+	"sigs.k8s.io/kubebuilder/v4/pkg/machinery"
+)
+
+var _ machinery.Template = &HelmChart{}
+
+// Chart scaffolds a file that defines the Helm chart structure
```
Comment (monteiro-renato) [inline pkg/plugins/optional/helm/v1alpha/scaffolds/internal/templates/chart.go:27]:
```suggestion
// HelmChart scaffolds a file that defines the Helm chart structure
```
nit: for consistency

File: pkg/plugins/optional/helm/v1alpha/scaffolds/init.go (hunk 17 above)
Comment (camilamacedo86) [reply pkg/plugins/optional/helm/v1alpha/scaffolds/init.go:131]: I opened a issue for that: https://github.com/kubernetes-sigs/kubebuilder/issues/4309

**Comment by camilamacedo86** [comment]
Hi @monteiro-renato 

Thank you a lot for your detailed review. 
I think I addressed all that you find so far.
Please, feel free to give another review.

**Review by monteiro-renato — APPROVED** [review]

File: pkg/plugins/optional/helm/v1alpha/scaffolds/init.go (hunk 17 above)
Comment (monteiro-renato) [reply pkg/plugins/optional/helm/v1alpha/scaffolds/init.go:131]: Kudos for opening an issue right away.

**Comment by camilamacedo86** [comment]
/hold cancel

We can merge now 
Patch release 4.3.1 is done

Thank you ALL for the reviews.

**Comment by camilamacedo86** [comment]
We had enough reviews and approvals for this initial alpha version
Anyway, anyone is more than welcome to push pull requests to improve the helm plugin
Your input and help is invaluable for us!

So, moving forward with this one !!!

**Comment by camilamacedo86** [comment]
I would like to thank you all for your help in the reviews 🥇 
That was great !!!
